tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -503,342 +503,546 @@ void emitNotToVectorReduction(OpBuilder& b, Location loc, Operation* root_op,
+  //       __syncthreads();
+  //       if (thread_row_index < stride) {
+  //         sdata[shm_trans_index] += sdata[shm_trans_index + stride];
+  //       }
+  //     }
+  //     __syncthreads();
+  for (int stride = kThreads_row / 2; stride > 1; stride /= 2) {
+    b.create<gpu::BarrierOp>(loc);
+    Value var_stride = b.create<arith::ConstantIndexOp>(loc, stride);
+    Value is_lt_stride = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::slt,
+                                                 thread_row_index, var_stride);
+    scf::IfOp if_stride_valid_op =
+        b.create<scf::IfOp>(loc, /*resultTypes*/ TypeRange{}, is_lt_stride,
+                            /*hasElseRegion*/ false);
+    if_stride_valid_op.getThenRegion().front().clear();
+    b.setInsertionPointToStart(&if_stride_valid_op.getThenRegion().front());
+    Value shm_offset =
+        b.create<arith::AddIOp>(loc, shm_trans_index, var_stride);
+    for (auto root_pair : llvm::enumerate(col_reduction_roots)) {
+      Operation* root_op = root_pair.value();
+      int idx = root_pair.index();
+      Value shm_op0 = b.create<memref::LoadOp>(loc, shared_mem_map[root_op],
+                                               shm_trans_index);
+      Value shm_op1 =
+          b.create<memref::LoadOp>(loc, shared_mem_map[root_op], shm_offset);
+      b.create<memref::StoreOp>(loc, (accum_factory[idx])(shm_op0, shm_op1),
+                                shared_mem_map[root_op], shm_trans_index);
     }
+    b.create<scf::YieldOp>(loc, ValueRange({}));
+    b.setInsertionPointAfter(if_stride_valid_op);
   }
+  b.create<gpu::BarrierOp>(loc);
 
-  b.setInsertionPointToEnd(parallel_op.getBody());
+  //     if (thread_row_index == 0) {
+  //       if (is_col_valid) {
+  //         accum = sdata[thread_col_index * num_threads_row];
+  //         atomicAdd(&global[col_index], accum);
+  //       }
+  //     }
+  //   }
+  // }
+  Value is_thread_row_zero = b.create<arith::CmpIOp>(
+      loc, arith::CmpIPredicate::eq, thread_row_index, zero);
+  Value is_valid = b.create<arith::AndIOp>(loc, is_thread_row_zero, is_lt_cols);
+  scf::IfOp if_valid_op =
+      b.create<scf::IfOp>(loc, /*resultTypes*/ TypeRange{}, is_valid,
+                          /*hasElseRegion*/ false);
+  if_valid_op.getThenRegion().front().clear();
+  b.setInsertionPointToStart(&if_valid_op.getThenRegion().front());
+  Value shm_offset1 = b.create<arith::AddIOp>(loc, shm_trans_index, one);
+  for (auto root_pair : llvm::enumerate(col_reduction_roots)) {
+    Operation* root_op = root_pair.value();
+    int idx = root_pair.index();
+    auto root_element_type = getLhloOpsElementType(root_op);
+    Value shm_op0 =
+        b.create<memref::LoadOp>(loc, shared_mem_map[root_op], shm_trans_index);
+    Value shm_op1 =
+        b.create<memref::LoadOp>(loc, shared_mem_map[root_op], shm_offset1);
+    Value partial_result = (accum_factory[idx])(shm_op0, shm_op1);
```

Comment:
Is this more effective to calculate the partial_result here than in `for (int stride = num_threads_row / 2; stride > 0; stride /= 2)`?

Reply:
Here I referred to the implementations of row reduction schedules and there is no performance gain. Now switched to `for (int stride = num_threads_row / 2; stride > 0; stride /= 2)`.